app/src/main/java/com/github/se/signify/model/common/user/FirestoreUserRepository.kt

```diff
@@ -530,4 +532,78 @@ class FirestoreUserRepository(
       }
     }
   }
+
+  override fun addPastChallenge(userId: String, challengeId: String) {
+    val userRef = db.collection(collectionPath).document(userId)
+    userRef.update("pastChallenges", FieldValue.arrayUnion(challengeId))
+  }
+
+  override fun getPastChallenges(
+      userId: String,
+      onSuccess: (List<Challenge>) -> Unit,
+      onFailure: (Exception) -> Unit
+  ) {
+    val userDocRef = db.collection(collectionPath).document(userId)
+
+    userDocRef
+        .get()
+        .addOnSuccessListener { document ->
+          if (!document.exists()) {
+            onSuccess(emptyList())
+            return@addOnSuccessListener
+          }
+
+          val challengeIds =
+              document.get("pastChallenges") as? List<*>
+                  ?: run {
+                    onSuccess(emptyList())
+                    return@addOnSuccessListener
+                  }
+
+          fetchChallengesByIds(challengeIds, onSuccess, onFailure)
+        }
+        .addOnFailureListener { e -> onFailure(e) }
+  }
+
+  override fun incrementField(userId: String, fieldName: String) {
+    val userRef = db.collection(collectionPath).document(userId)
+    userRef.update(fieldName, FieldValue.increment(1))
+  }
+
+  override fun updateUserField(
+      userId: String,
+      fieldName: String,
+      value: Any,
+      onSuccess: () -> Unit,
+      onFailure: (Exception) -> Unit
+  ) {
+    val userRef = db.collection(collectionPath).document(userId)
+    userRef
+        .update(fieldName, value)
+        .addOnSuccessListener { onSuccess() }
+        .addOnFailureListener { e -> onFailure(e) }
+  }
+
+  override suspend fun getChallengesCompleted(userId: String): Int {
+    return getUserField(userId, "challengesCompleted")
+  }
+
+  override suspend fun getChallengesCreated(userId: String): Int {
+    return getUserField(userId, "challengesCreated")
+  }
+
+  override suspend fun getChallengesWon(userId: String): Int {
+    return getUserField(userId, "challengesWon")
+  }
+
```

Comment:
You should add onSuccess and onFailure parameters to be consistent with the other functions of the Repository and to ensure the security of the application.

Reply:
Nice catch! I did it.